### ISSUE TITLE
OOM killer triggered on CI

### ISSUE THREAD

**Issue by boris-kolpackov** [build2-packaging/glbinding#1]
I've noticed that a few CI jobs (linux, GCC 12.1, optimized) got killed due to running out of memory. The VM in question has 16GB RAM but it does compile on 12 cores in parallel.

Just checking if this is something expected (i.e., the library has unusually large TUs or some such).

**Comment by lyrahgames** [comment]
Thanks for the notice. I tried to update the package to properly use test packages and the configuration features from build2 0.15.0. And indeed, since turning off my naive attempt for LTO on Linux, GCC uses up an enormous amount of memory during compilation of the glbinding sources. The process got killed on my machines as well and I had to run it twice to finish the job. Actually, I do not think that this behavior was intended by the developers. That is one of the two reasons (the other concerning problematic compilation issues with Clang and GCC on Windows) that I have not published the new version 3.3.0 of the wrapper package to cppget.org. Presumably, it will take me some time to resolve this because I, currently, need to finish my thesis. If you have some suggestions, I would appreciate them very much.

**Comment by boris-kolpackov** [comment]
Hm, I am not sure what's going on here. I could understand if turning LTO on would cause GCC to run out of memory. But running out of memory after turning it off is surprising. But maybe I am missing something.

BTW, we do have some support for LTO in GCC and Clang:  https://github.com/build2/build2/blob/master/NEWS#L702-L708 So maybe you can leave it enabled if it solved the problem?

**Comment by lyrahgames** [comment]
> Hm, I am not sure what's going on here. I could understand if turning LTO on would cause GCC to run out of memory. But running out of memory after turning it off is surprising. But maybe I am missing something.

No, you are right. I did not understand it, either. So, that was the reason, I wanted to take a deeper look into this because the developers made LTO optional and disabled it by default. Using `-flto=auto` with GCC worked only for dynamic libraries while still using up to 70 threads or so. I was not able to build and use the static library with LTO enabled.

> BTW, we do have some support for LTO in GCC and Clang: https://github.com/build2/build2/blob/master/NEWS#L702-L708 So maybe you can leave it enabled if it solved the problem?

Oh, I am surprised and very sorry. I didn't know that and must have overread it. But I will definitely take a look into this.

**Comment by boris-kolpackov** [comment]
FYI, still running out of memory on `linux_debian_12-bindist-gcc_12`. The VM has 16GB or RAM and compiles on 8 cores.

**Comment by lyrahgames** [comment]
Yeah, sorry. It is hard to focus on this issue, as I had to deal with several other problems in this library.
For now, I completely disabled LTO. I have to admit that this package is driving me nuts. ^^

**Comment by boris-kolpackov** [comment]
> For now, I completely disabled LTO.

```
default.coptions += -fPIC -fvisibility=hidden -flto=auto
default.loptions += -flto=auto
```

Thinking about this, I believe it's actually wrong to enable `LTO` at the `buildfile` level in general, just as it would be wrong to, say, add `-O3` there. For background, see: https://github.com/build2/HOWTO/blob/master/entries/compile-options-in-buildfile.md

In other words, it's the user's decision to make whether to optimize, and if so, whether to also do LTO. If you do want to CI with LTO (in addition to non-LTO), then the correct place to enable it would be in package build configuration: https://build2.org/stage/bpkg/doc/build2-package-manager-manual.xhtml#manifest-package-build-config

> I have to admit that this package is driving me nuts.

Yeah, I've noticed all the graphics libraries tend to be quite nutty for some reason...

**Comment by lyrahgames** [comment]
> > For now, I completely disabled LTO.
> 
> ```
> default.coptions += -fPIC -fvisibility=hidden -flto=auto
> default.loptions += -flto=auto
> ```

These flags have been commented out for the last few commits.
I had to change the handling of the export header into a more build2-like way and removed these options for testing.
See: 5feb3d2f32775a24f25d2cf6782b0911fdd1e1a1

> Thinking about this, I believe it's actually wrong to enable `LTO` at the `buildfile` level in general, just as it would be wrong to, say, add `-O3` there. For background, see: https://github.com/build2/HOWTO/blob/master/entries/compile-options-in-buildfile.md
> 
> In other words, it's the user's decision to make whether to optimize, and if so, whether to also do LTO. If you do want to CI with LTO (in addition to non-LTO), then the correct place to enable it would be in package build configuration: https://build2.org/stage/bpkg/doc/build2-package-manager-manual.xhtml#manifest-package-build-config

That was my plan as soon as soon as the compilation and tests would run stable under Linux, MacOS, and Windows (at least with MSVC) without enabling LTO. Especially, LTO seems to require at least `-O2` or `-O3` as the debug build already runs out of memory on my machine. A build config would be perfect for that.
By the way, is it possible to initialize a package with `bdep` according to one of its prepared and named build configs?

**Comment by boris-kolpackov** [comment]
> By the way, is it possible to initialize a package with bdep according to one of its prepared and named build configs?

You mean the ones specified in the manifest with `*-build-config`? If so, not at the moment but that's an interesting idea, thanks: https://github.com/build2/build2/issues/298

**Comment by boris-kolpackov** [comment]
FYI, the submitted `3.3.0` package triggers OOM sometimes, appears to be during compilation (Linux, GCC, -O3):

```
Out of memory: Killed process 1857 (cc1plus) ... pgtables:13572kB ...
```

I am not sure which file(s) exactly triggers this, but I am pretty sure it happens during compilation (don't see any linker command lines yet).

In the latest staged toolchain I've added a mechanism for dealing with this: https://github.com/build2/build2/commit/82ff2e7df2243b679aadbc6cc120a2b5f7ee73b3

Could you try to use it to prevent OOM kills here? Specifically:

1. Identify which source files require a large amount of RAM by compiling serially (`-s`) and watching the amount of RAM used by `cc1plus` for each compilation.

2. Serialize compilation of all such files using the above mechanism.

If this works out, I think you can just publish a revision with this change: it will have no effect until 0.17.0 is released, but once that happens, this will presumably be taken care of.

**Comment by lyrahgames** [comment]
Thanks for taking a look into this one. Sounds like a good solution. I will try to do this as soon as possible (hopefully next week) but it will take some time.

**Comment by boris-kolpackov** [comment]
Ok, thanks, let me know how it goes.

In the meantime, I've dropped the current submission from queue.cppget.org since it keeps hanging our CI bots (it looks like when OOM killer kills the `cc1plus` process, the GCC driver does not exist which in turn hangs the build process).

**Comment by boris-kolpackov** [comment]
`0.17.0` is out so perhaps it's time to publish the revision with the fix?

Also, I've added a HOWTO for this problem to the packaging guide:

https://build2.org/stage/build2-toolchain/doc/build2-toolchain-packaging.xhtml#howto-out-of-ram

**Comment by lyrahgames** [comment]
Thank you for the HOWTO entry and for the great work on version 0.17.0. I have already added the change and prepared the revision a few weeks ago. So far I haven't published it because I was waiting for answer to my email to you on how to further proceed with more complicated revisions. However, I know that this small change completely satisfies the rules of backwards compatible revision. I will publish it as soon as I am back in Germany.

**Comment by boris-kolpackov** [comment]
Ok, thanks, that sounds good. I see your emails but I haven't gotten to them yet in the process of clearing my backlog. It may still take me some weeks, unfortunately.

**Comment by lyrahgames** [comment]
Yes, no rush. I assumed so as you only released the new version a few days ago.

**Comment by lyrahgames** [comment]
I published the new revision and hope everything is ok with it.

**Comment by boris-kolpackov** [comment]
Yes, the queued packages seem to build fine now, thanks.

I tried to quickly review and publish the packages by examining the difference compared to the previous version, but it seems there are a lot of changes (and new packages). Even with the cursory look I already see some questionable things so this will require a proper, from-scratch re-review. Not sure when I will be able to get to it, sorry.
